editor/src/components/canvas/ui-jsx-canvas-renderer/ui-jsx-canvas-component-renderer.tsx

```diff
@@ -49,10 +59,24 @@ export function createComponentRendererComponent(params: {
 
     const scenePath: ScenePath | null = TP.isScenePath(scenePathAny) ? scenePathAny : null
 
-    const { current: mutableContext } = React.useContext(MutableUtopiaContext)
-    const utopiaJsxComponent = useContextSelector(RerenderUtopiaContext, (c) =>
-      c.topLevelElements.get(params.topLevelElementName),
-    )
+    const mutableContext = params.mutableContextRef.current
+
+    const utopiaJsxComponent = useEditorState((store) => {
+      const projectFile = getFileForName(params.filePath, store.editor)
+      if (isTextFile(projectFile) && isParseSuccess(projectFile.fileContents.parsed)) {
+        return (
+          mapDropNulls((elem) => {
+            if (isUtopiaJSXComponent(elem) && elem.name === params.topLevelElementName) {
+              return elem
+            } else {
+              return null
+            }
+          }, projectFile.fileContents.parsed.topLevelElements)[0] ?? null
+        )
+      } else {
+        return null
+      }
+    }, 'createComponentRendererComponent utopiaJsxComponent')
```

Comment:
This block will run every time the component's props change, right? Could we move the block outside of the component and into the `createComponentRendererComponent` (meaning it would have to become a hook) or into the `ui-jsx-canvas`?

Reply:
we need to re-render the component if the related topLevelElement changes. but we want to avoid re-mounting the component, so it needs to be a hook and it needs to live inside the component renderer component. it's not a problem that this hook is called every time the component is rendered, since the useEditorState hooks are designed to be cheap and fast